sky/utils/controller_utils.py

```diff
@@ -305,7 +316,39 @@ def get_controller_resources(
                     err=f'Expected exactly one resource, got '
                     f'{len(controller_resources)} resources: '
                     f'{controller_resources}'))
-    return list(controller_resources)[0]
+    controller_resources_to_use: resources.Resources = list(
+        controller_resources)[0]
+
+    controller_exist = (global_user_state.get_cluster_from_name(controller_name)
+                        is not None)
+    if controller_exist or controller_resources_to_use.cloud is not None:
+        return {controller_resources_to_use}
+
+    # If the controller and replicas are from the same cloud, it should
+    # provide better connectivity. We will let the controller choose from
+    # the clouds of the resources if the controller does not exist.
+    # TODO(tian): Consider respecting the regions/zones specified for the
+    # resources as well.
+    requested_clouds: Set['clouds.Cloud'] = set()
+    for res in task_resources:
+        # cloud is an object and will not be able to be distinguished by set.
+        # Here we manually check if the cloud is in the set.
+        if res.cloud is not None and not clouds.cloud_in_iterable(
```

Comment:
I am thinking maybe if any of the `res.cloud is None`, we should allow the controller to be launched on any clouds, because it is possible that a user may specify resources to be the following, we should allow controller to start on aws or gcp, etc.
```
resources:
  any_of:
    - accelerators: L4
    - cloud: runpod
       accelerators: A40
```

Reply:
Good point! Added. Thanks!